src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,173 @@
+    size_t index = 1;
+    for (auto& statement : context->pred_arg()) {
+        args_.push_back(std::move(statement->ID()->getText()));
+        index += 1;
+    }
+
+    if (context->NEG() == nullptr) {
+        sign_ = datalog::Sign::kPositive;
+    }
+    else {
+        sign_ = datalog::Sign::kNegated;
+    }
+    return datalog::Predicate(name_, args_, sign_);
+
+}
+
+template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::VerbphraseContext, T>::value>>
+T* instanceofVerbPhrase(AuthLogicParser::VerbphraseContext* context) {
+     return dynamic_cast<T*>(context);
+}
+
+absl::StatusOr<BaseFact> construct_verbphrase(Principal left_principal_, AuthLogicParser::VerbphraseContext* context) {
+    if (auto* ctx = instanceofVerbPhrase<AuthLogicParser::PredphraseContext>(context)) {
+
+        datalog::Predicate predicate_ = construct_predicate(ctx->predicate());   //PredicateContext
+        return BaseFact(Attribute(left_principal_, datalog::Predicate(predicate_)));
+    }
+    else if (auto* ctx = instanceofVerbPhrase<AuthLogicParser::ActsAsPhraseContext>(context)) {
+        Principal right_principal_ = construct_principal(ctx->principal());    //ActsAsPhraseContext
+        return BaseFact(CanActAs(left_principal_, right_principal_));
+    }
+
+    return absl::InvalidArgumentError("Invalid context to construct_verbphrase");
+}
+
+
+template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::FlatFactContext, T>::value>>
+T* instanceofFlatFactContext(AuthLogicParser::FlatFactContext* context) {
+     return dynamic_cast<T*>(context);
+}
+
+absl::StatusOr<BaseFact> construct_flatFact(AuthLogicParser::FlatFactContext* context) {
+    if (auto* ctx = instanceofFlatFactContext<AuthLogicParser::PrinFactContext>(context)) {                                              
+        Principal principal_ = construct_principal(ctx->principal());
+        return construct_verbphrase(principal_, ctx->verbphrase());
+    }
+    else if (auto* ctx = instanceofFlatFactContext<AuthLogicParser::PredFactContext>(context)) {
+        datalog::Predicate predicate_ = construct_predicate(ctx->predicate());   //predFactContext
+        return BaseFact(predicate_);
+    }
+
+    return absl::InvalidArgumentError("Invalid context to construct_flatFact");
+}
+
+template<typename T, typename = std::enable_if_t<std::is_base_of<AuthLogicParser::FactContext, T>::value>>
+T* instanceofFactContext(AuthLogicParser::FactContext* context) {
+     return dynamic_cast<T*>(context);
+}
+
+absl::StatusOr<Fact> construct_fact(AuthLogicParser::FactContext* context) {
+    if (auto* flat_fact_fact_ctx = instanceofFactContext<AuthLogicParser::FlatFactFactContext>(context)) {  
+        absl::StatusOr<BaseFact> result = construct_flatFact(flat_fact_fact_ctx->flatFact());
+        if (result.ok()) {
+            BaseFact basefact = *result;                               //FlatFactFactContext
+            return Fact(basefact);
+        }
```

Comment:
The following might be a better pattern?
```suggestion
        if (!result.ok())  return result; // Break on error.
        BaseFact basefact = *result;                               //FlatFactFactContext
        return Fact(basefact);
```